promql/engine.go

```diff
@@ -0,0 +1,1262 @@
+	for _, ls := range lhs {
+		hash := hashForMetric(ls.Metric.Metric, matching.On)
+		// Any lhs sample we encounter in an OR operation belongs to the result.
+		if op == itemLOR {
+			ls.Metric = resultMetric(op, ls, nil, matching)
+			result = append(result, ls)
+			added[hash] = nil // Ensure matching rhs sample is not added later.
+			continue
+		}
+
+		rs, found := rm[hash] // Look for a match in the rhs vector.
+		if !found {
+			continue
+		}
+		var value clientmodel.SampleValue
+		var keep bool
+
+		if op == itemLAND {
+			value = ls.Value
+			keep = true
+		} else {
+			if _, exists := added[hash]; matching.Card == CardOneToOne && exists {
+				// Many-to-one matching must be explicit.
+				ev.fatalf("many-to-one matching must be explicit")
+			}
+			// Account for potentially swapped sidedness.
+			vl, vr := ls.Value, rs.Value
+			if matching.Card == CardOneToMany {
+				vl, vr = vr, vl
+			}
+			value, keep = ev.vectorElemBinop(op, vl, vr)
+		}
+
+		if keep {
+			metric := resultMetric(op, ls, rs, matching)
+			// Check if the same label set has been added for a many-to-one matching before.
+			if matching.Card == CardManyToOne || matching.Card == CardOneToMany {
+				insHash := clientmodel.SignatureForLabels(metric.Metric, matching.Include)
+				if ihs, exists := added[hash]; exists {
+					for _, ih := range ihs {
+						if ih == insHash {
+							ev.fatalf("metric with label set has already been matched")
+						}
+					}
+					added[hash] = append(ihs, insHash)
+				} else {
+					added[hash] = []uint64{insHash}
+				}
+			}
+			ns := &Sample{
+				Metric:    metric,
+				Value:     value,
+				Timestamp: ev.Timestamp,
+			}
+			result = append(result, ns)
+			added[hash] = added[hash] // Set existance to true.
+		}
+	}
+
+	// Add all remaining samples in the rhs in an OR operation if they
+	// have not been matched up with a lhs sample.
+	if op == itemLOR {
+		for hash, rs := range rm {
+			if _, exists := added[hash]; !exists {
+				rs.Metric = resultMetric(op, rs, nil, matching)
+				result = append(result, rs)
+			}
+		}
+	}
+	return result
+}
+
+// vectorScalarBinop evaluates a binary operation between a vector and a scalar.
+func (ev *evaluator) vectorScalarBinop(op itemType, lhs Vector, rhs *Scalar, swap bool) Vector {
+	vector := make(Vector, 0, len(lhs))
+
+	for _, lhsSample := range lhs {
+		lv, rv := lhsSample.Value, rhs.Value
+		// lhs always contains the vector. If the original position was different
+		// swap for calculating the value.
+		if swap {
+			lv, rv = rv, lv
+		}
+		value, keep := ev.vectorElemBinop(op, lv, rv)
+		if keep {
+			lhsSample.Value = value
+			if shouldDropMetricName(op) {
+				lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
+			}
+			vector = append(vector, lhsSample)
+		}
+	}
+	return vector
+}
+
+// scalarBinop evaluates a binary operation between two scalars.
+func (ev *evaluator) scalarBinop(op itemType, lhs, rhs clientmodel.SampleValue) clientmodel.SampleValue {
```

Comment:
This and the method below are not really methods because they don't use `ev`. Functionize them?